### DISCUSSION TITLE
Proposal: syntactic sugar for initializing Span<char> by stackalloc with string literal

### DISCUSSION THREAD

**Discussion by acid-chicken** [dotnet/csharplang#8761] [Language Ideas]
# Summary

Provide a syntactic sugar which initializes Span<char> and ReadOnlySpan<char> with stackalloc keyword by string literal.

# Design

```csharp
Span<char> span = stackalloc [] "initial characters";
```

will be converted to

```csharp
Span<char> span = stackalloc [] { 'i', 'n', 'i', 't', 'i', 'a', 'l', ' ', 'c', 'h', 'a', 'r', 'a', 'c', 't', 'e', 'r', 's' };
```

### Syntax Style

```csharp
Span<char> span = stackalloc [] "initial characters";
```

or

```csharp
Span<char> span = stackalloc "initial characters";
```

or

```csharp
Span<char> span = stackalloc char[] "initial characters";
/* NEED TO SPECIFY TYPE STRICTLY */
```

or other?

### Also Support `char[]`

Also adding syntactic suger for `char[]` initializer?

**Comment by svick** [comment]
Why? Are there any cases where memory consumed by string literals is an issue?

**Comment by jnm2** [comment]
@svick I read this as wanting to initialize the space from a string literal but then using it as scratch space.

**Comment by svick** [comment]
@jnm2 

Oh, so the equivalent of the following code?

```c#
string s = "initial characters";
Span<char> span = stackalloc char[s.Length];
s.AsSpan().CopyTo(span);
```

Is that actually so common that it needs syntactic sugar?

**Comment by AraHaan** [comment]
I do not think so. I did not even know about Span in c#.

**Comment by jnm2** [comment]
@svick That's what I understood @acid-chicken to mean. I haven't run into this pattern before, though I have been in situations where I'd copy a literal to the beginning of a longer buffer.

If the stack space isn't going to be mutated after copying the literal into it, I'd just get the pointer to the literal directly without allocating stack space and doing the copy (or use ReadOnlySpan).
